internal/magic/text.go

```diff
@@ -13,6 +13,7 @@ import (
 var (
 	// HTML matches a Hypertext Markup Language file.
 	HTML = markup(
+		append([]byte{0xEF, 0xBB, 0xBF}, []byte("<!DOCTYPE HTML")...),
```

Comment:
This line is not needed anymore if you trim the BOM.

Reply:
Hi @gabriel-vasile  the reason for restoring the BOM is the following: the charset encoding specified in meta may be erroneous or missing - having the BOM enables the document to be reliably classified as UTF-8. I base this on the following [statement](https://www.w3.org/International/questions/qa-byte-order-mark) by the W3C : 

`In HTML5 browsers are required to recognize the UTF-8 BOM and use it to detect the encoding of the page, and recent versions of major browsers handle the BOM as expected when used for UTF-8 encoded pages.`

So my reading of it is that if the BOM is present in an HTML5 page it should be used to classify it as UTF8, possibly overriding the charset declaration. I think the potential performance hit from a possible copy is quite low (but of course we can benchmark it to make sure).

Reply:
So you're saying BOM charset should have precedence over what is in `<meta charset="foobar">`.

Looking at https://html.spec.whatwg.org/multipage/parsing.html#determining-the-character-encoding, it seems like you are right, and BOM takes over meta. Still, there is no need to preserve the BOM, because [charset detection](https://github.com/gabriel-vasile/mimetype/blob/71f0e2b03be1cdff67bc5447c554d05daf9b19b4/mime.go#L101-L112) receives a copy of the original file data. The problem is [charset.FromHTML](https://github.com/gabriel-vasile/mimetype/blob/71f0e2b03be1cdff67bc5447c554d05daf9b19b4/internal/charset/charset.go#L160-L167) looks first in `<meta>` and then at BOM.

Reply:
Ah I didn't realize that charset detection worked on a copy of the `sig` data in that case you are right we can strip the BOM - I was just about to post that we could avoid the re-alloc and copy by doing something like:
```go
// trimLWS trims whitespace from beginning of the input.
func trimLWS(in []byte) []byte {
	var firstNonWS int

	if bytes.HasPrefix(in, []byte{0xEF, 0xBB, 0xBF}) {
		firstNonWS = 3
		for ; firstNonWS < len(in) && isWS(in[firstNonWS]); firstNonWS++ {
		}
		copy(in[firstNonWS-3:], []byte{0xEF, 0xBB, 0xBF})

	} else {
		for ; firstNonWS < len(in) && isWS(in[firstNonWS]); firstNonWS++ {
		}
	}

	return in[firstNonWS:]
}
```
So let's strip the BOM in `markup` and modify the charset detection to ensure the BOM if present is used first - agreed?

Reply:
> So let's strip the BOM in `markup` and modify the charset detection to ensure the BOM if present is used first - agreed?

Yes, agree.